awx/api/metadata.py

```diff
@@ -252,6 +252,16 @@ def determine_metadata(self, request, view):
         if getattr(view, 'related_search_fields', None):
             metadata['related_search_fields'] = view.related_search_fields
 
+        # include role names in metadata
+        roles = []
+        model = getattr(view, 'model', None)
+        if model:
+            for field in model._meta.get_fields():
+                if type(field) is ImplicitRoleField:
+                    roles.append(field.name)
```

Comment:
This loop is simple and the most obvious solution to do this particular task. But to be consistent with our current code base, we should instead loop over `__implicit_role_fields` which is set by the field metaclass hooks.

https://github.com/ansible/awx/blob/320284267cd5de9326ff34fbcd09e075391069d0/awx/main/fields.py#L240-L242

If this doesn't really make sense, then I think we should aim to get rid of `__implicit_role_fields`, because it's one more thing in the field code just waiting to mess up our migrations.

Reply:
I stole this code almost verbatim from here:

https://github.com/ansible/awx/blob/devel/awx/api/serializers.py#L426